contracts/AMMWrapper.sol

```diff
@@ -336,9 +334,17 @@ contract AMMWrapper is IAMMWrapper, ReentrancyGuard, BaseLibEIP712, SignatureVal
         TxMetaData memory _txMetaData,
         InternalTxData memory _internalTxData
     ) internal returns (uint256 settleAmount) {
-        // settleAmount = receivedAmount - fee
-        uint256 fee = _order.makerAssetAmount.mul(feeFactor).div(LibConstant.BPS_MAX);
-        settleAmount = _txMetaData.receivedAmount.sub(fee);
+        if (_txMetaData.receivedAmount > _order.makerAssetAmount) {
+            // shouldCollectFee = ((receivedAmount - makerAssetAmount) / receivedAmount) > (feeFactor / 10000)
+            bool shouldCollectFee = _txMetaData.receivedAmount.sub(_order.makerAssetAmount).mul(LibConstant.BPS_MAX) >
+                feeFactor.mul(_txMetaData.receivedAmount);
+            if (shouldCollectFee) {
+                // settleAmount = receivedAmount * (1 - feeFactor) / 10000
+                settleAmount = _txMetaData.receivedAmount.mul(LibConstant.BPS_MAX.sub(feeFactor)).div(LibConstant.BPS_MAX);
+            } else {
+                settleAmount = _order.makerAssetAmount;
+            }
+        }
```

Comment:
Use an example to clarify the logic, suppose fee factor is 5% and make amount is 100, then

(1) swap out 99 -> revert
(2) swap out 103 -> maker takes 100, our AMM takes 3
(3) swap out 110 -> maker takes 104.5, our AMM takes 5.5

So when swapped out surplus is less than our fee factor, we would still collect fee, but it would be less than fee factor set on contract, and maker would receive exactly the maker amount. For maker to receive more than maker amount, the swapped out surplus must be more than fee factor.

Is this understanding correct?

Reply:
I think we should adjust the frontend to show the amount after fee is charged instead of doing the calculation and determine if we should charge fee in contract, i.e., we should charge as long as the AMM swap succeeded.

Because if the `makerAssetAmount` is supplied not by us but by user, user can fine tune the `makerAssetAmount` so we can never charge fee from him.

Reply:
Sounds reasonable.